File: packages/primitives/src/query/validate.test.ts
// §6's rules, in both polarities.
//
// "The validator is mutation-tested in both polarities: each rule must reject
// the thing it names AND accept the nearest legal query, because an over-strict
// guard is the same bug pointed the other way." (§6.) The conformance file
// carries a twin for every failing case; these add the ones a port can get
// wrong on its own — chiefly the caps whose units differ, and the suggestions,
// which the fixture deliberately does not pin so wording can improve.

import { describe, expect, it } from 'vitest';

import { check } from './index';
import type { QueryError } from './errors';
import { parse } from './parser';
import { LADDER, newTestCatalog } from './test-catalog';
import { defaultOptions, validate, withLadder } from './validate';
import type { Options } from './validate';

const cat = newTestCatalog();

function opts(over: Partial<Options> = {}): Options {
  return { ...withLadder(defaultOptions(), LADDER), ...over };
}

function codes(src: string, target = 'asset', over: Partial<Options> = {}): string[] {
  const out = check(src, target, cat, opts(over));
  return out.ok ? [] : out.errors.map((e) => e.code);
}

function firstError(src: string, target = 'asset', over: Partial<Options> = {}): QueryError {
  const out = check(src, target, cat, opts(over));
  if (out.ok) throw new Error(`expected ${JSON.stringify(src)} to fail`);
  return out.errors[0];
}

/** Every rule, with the nearest query that must still be accepted. */
const BOTH_POLARITIES: { rule: string; bad: string; code: string; ok: string; over?: Partial<Options> }[] = [
  { rule: 'unknown_field', bad: 'hostnaem:web-1', code: 'unknown_field', ok: 'hostname:web-1' },
  {
    rule: 'unknown_field (namespace key)',
    bad: 'attr.not_registered:1',
    code: 'unknown_field',
    ok: 'attr.os_version:1',
  },
  {
    rule: 'operator_not_allowed',
    bad: 'environment < production',
    code: 'operator_not_allowed',
    ok: 'environment:production',
  },
  { rule: 'type_mismatch', bad: 'risk >= 70', code: 'type_mismatch', ok: 'risk_score >= 70' },
  { rule: 'unknown_value', bad: 'status:monitorng', code: 'unknown_value', ok: 'status:monitoring' },
  {
    rule: 'unknown_value (class)',
    bad: 'class:serverr',
    code: 'unknown_value',
    ok: 'class:server',
  },
  {
    rule: 'depth_exceeded',
    bad: 'depends_on(4):(class:server)',
    code: 'depth_exceeded',
    ok: 'depends_on(3):(class:server)',
  },
  {
    rule: 'regex_invalid',
    bad: 'hostname ~ "\\\\bweb"',
    code: 'regex_invalid',
    ok: 'hostname ~ "^web"',
  },
  {
    rule: 'query_too_long',
    bad: 'environment:production and class:server',
    code: 'query_too_long',
    ok: 'class:server',
    over: { maxBytes: 20 },
  },
  {
    rule: 'too_many_clauses',
    bad: 'class:server and environment:production and status:monitoring',
    code: 'too_many_clauses',
    ok: 'class:server and environment:production',
    over: { maxLeaves: 2 },
  },
  {
    rule: 'untranslatable (free text off-target)',
    bad: 'payroll',
    code: 'untranslatable',
    ok: 'subject_dn:payroll',
  },
];

describe('§6, in both polarities', () => {
  for (const c of BOTH_POLARITIES) {
    it(`${c.rule}: rejects the thing it names`, () => {
      const target = c.rule.includes('off-target') ? 'certificate' : 'asset';
      expect(codes(c.bad, target, c.over)).toContain(c.code);
    });
    it(`${c.rule}: accepts the nearest legal query`, () => {
      const target = c.rule.includes('off-target') ? 'certificate' : 'asset';
      expect(codes(c.ok, target, c.over)).toEqual([]);
    });
  }
});

describe('the caps, and the units they are measured in', () => {
  // The fixture file is pure ASCII, so it cannot tell a byte cap from a
  // character cap. §6 says the query cap is BYTES and the regex cap is
  // CHARACTERS, "on purpose" — a port that used `String.length` for both would
  // pass all 228 fixtures and then disagree with the server on the first
  // accented hostname.
  it('measures the query cap in bytes, not code units', () => {
    // 10 accented characters: 10 code units, 20 UTF-8 bytes.
    const q = 'display_name:"' + 'é'.repeat(10) + '"';
    expect(q.length).toBe(25);
    expect(codes(q, 'asset', { maxBytes: 30 })).toEqual(['query_too_long']);
    expect(codes(q, 'asset', { maxBytes: 40 })).toEqual([]);
  });

  it('measures the regex cap in code points, not bytes', () => {
    // Eight accented characters is eight CHARACTERS and sixteen bytes.
    const q = 'hostname ~ "' + 'é'.repeat(8) + '"';
    expect(codes(q, 'asset', { maxRegexLen: 8 })).toEqual([]);
    expect(codes(q, 'asset', { maxRegexLen: 7 })).toEqual(['regex_invalid']);
  });

  it('caps the values in one list', () => {
    // `business_unit`, not `environment`: the latter is a closed set since
    // §12 A1, and arbitrary values there would fail for a second, unrelated
    // reason — which would make this test pass for the wrong one.
    expect(codes('business_unit in (a, b, c)', 'asset', { maxInValues: 2 })).toEqual([
      'too_many_clauses',
    ]);
    expect(codes('business_unit in (a, b)', 'asset', { maxInValues: 2 })).toEqual([]);
  });

  it('caps parenthesis nesting on what was WRITTEN, not what survives formatting', () => {
    // Canonical form drops redundant parentheses, so the AST no longer shows
    // what the user typed; the depth is measured at parse time for that reason.
    expect(codes('(((class:server)))', 'asset', { maxParenDepth: 2 })).toEqual([
      'too_many_clauses',
    ]);
    expect(codes('((class:server))', 'asset', { maxParenDepth: 2 })).toEqual([]);
  });

  it('applies the hard traversal maximum over any configured budget (§5.6)', () => {
    expect(codes('depends_on(7):(class:server)', 'asset', { traversalBudget: 10 })).toEqual([
      'depth_exceeded',
    ]);
    expect(codes('depends_on(6):(class:server)', 'asset', { traversalBudget: 10 })).toEqual([]);
  });

  it('sums declared depths along the deepest nested path', () => {
    // 2 + 1 = 3 is exactly the default budget; 2 + 2 is not.
    expect(codes('depends_on(2):(hosted_on:(class:hypervisor))')).toEqual([]);
    expect(codes('depends_on(2):(hosted_on(2):(class:hypervisor))')).toEqual(['depth_exceeded']);
  });
});

describe('the regex dialect (§13 A6)', () => {
  // Each of these is inside RE2 and outside what Postgres does with it, or the
  // other way round. `\b` is the one that matters most: it compiles as RE2,
  // validates, reaches Postgres — where it is a BACKSPACE — and returns the
  // wrong rows with no error anywhere.
  const outside = [
    ['\\\\bweb\\\\b', 'word boundary is a backspace in ARE'],
    ['web(?i)01', '(?i) is a prefix-only flag'],
    ['(?<name>x)', 'named groups'],
    ['x\\\\z', 'anchors outside the subset'],
    ['\\\\p{L}', 'unicode classes'],
    ['a+?', 'non-greedy quantifiers mean different things'],
    ['[[:alpha:]]', 'POSIX classes inside a class'],
    ['a{256}', 'past Postgres DUPMAX'],
  ];
  for (const [pattern, why] of outside) {
    it(`rejects ${pattern} — ${why}`, () => {
      expect(codes(`hostname ~ "${pattern}"`)).toEqual(['regex_invalid']);
    });
  }

  const inside = ['^web[0-9]{2}$', '(?i)web', 'a{255}', '(a|b)+', '\\\\d+\\\\.\\\\d+', '[^a-z]'];
  for (const pattern of inside) {
    it(`accepts ${pattern}`, () => {
      expect(codes(`hostname ~ "${pattern}"`)).toEqual([]);
    });
  }

  it('caps the repetition PRODUCT across nesting, separately from a single count', () => {
    // Each bound is legal on its own; Postgres calls their product "too
    // complex", which reaches a user as a 500 rather than a diagnostic.
    expect(codes('hostname ~ "(a{100}){100}"')).toEqual(['regex_invalid']);
    expect(codes('hostname ~ "(a{10}){100}"')).toEqual([]);
  });
});

describe('three-valued logic, as the validator can see it (§5.2)', () => {
  // `not_assessed` is the absence of a score, not a rung of the ladder. Before
  // §13 A2, `risk < not_assessed` validated and then returned every unassessed
  // asset, and `risk:[not_assessed to high]` validated and then failed
  // `untranslatable` — a code §6 reserves for a parser bug.
  it('allows not_assessed only with : = and !=', () => {
    expect(codes('risk:not_assessed')).toEqual([]);
    expect(codes('risk != not_assessed')).toEqual([]);
    expect(codes('risk < not_assessed')).toEqual(['operator_not_allowed']);
    expect(codes('risk:[not_assessed to high]')).toEqual(['operator_not_allowed']);
  });

  it('refuses not_assessed on a band that records no coverage', () => {
    // `finding.severity` has no assessedBy column, so "nobody assessed it" is
    // not a question that field can answer.
    expect(codes('severity:not_assessed', 'finding')).toEqual(['unknown_value']);
    expect(codes('severity:critical', 'finding')).toEqual([]);
  });

  it('refuses a version field with no normalised sort key', () => {
    // §5.5 forbids a lexical fallback outright, so a version field whose
    // accessor names no sort column is untranslatable rather than compared as
    // a string. Every version field in the catalogue HAS one, which is what the
    // passing half proves.
    expect(codes('version:3.0', 'software_install')).toEqual([]);
    const res = parse('version:3.0');
    expect(res.ok).toBe(true);
    if (!res.ok) return;
    const stripped = {
      ...cat,
      targets: () => cat.targets(),
      fields: (t: string) => cat.fields(t),
      relationshipNames: () => cat.relationshipNames(),
      classExists: (k: string) => cat.classExists(k),
      enumValues: () => null,
      resolve: (t: string, p: string[]) => {
        const out = cat.resolve(t, p);
        if (!out.ok) return out;
        return { ok: true as const, field: { ...out.field, accessor: { ...out.field.accessor, sortColumn: '' } } };
      },
    };
    const out = validate(res.value, 'software_install', stripped, opts());
    expect(out.ok).toBe(false);
    if (!out.ok) expect(out.errors.map((e) => e.code)).toEqual(['untranslatable']);
  });

  it('refuses a band field when no ladder was supplied', () => {
    // Without a ladder a band is untranslatable — refused, not guessed at.
    const noLadder = defaultOptions();
    expect(check('risk:high', 'asset', cat, noLadder)).toMatchObject({
      ok: false,
      errors: [{ code: 'untranslatable' }],
    });
    expect(check('risk_score >= 70', 'asset', cat, noLadder).ok).toBe(true);
  });
});

describe('reachability of a target (§5.1, §5.6)', () => {
  it('refuses a collection the target cannot reach, and names the ones it can', () => {
    const e = firstError('finding:(kind:stale)', 'endpoint');
    expect(e.code).toBe('untranslatable');
    expect(e.suggestion).toContain('asset');
    expect(codes('asset:(class:server)', 'endpoint')).toEqual([]);
  });

  it('refuses traversal off an asset, and says how to wrap it', () => {
    const e = firstError('depends_on:(class:server)', 'certificate');
    expect(e.code).toBe('untranslatable');
    expect(e.suggestion).toContain('asset:(depends_on:(…))');
    expect(codes('asset:(depends_on:(class:server))', 'certificate')).toEqual([]);
  });

  it('refuses a reverse label in the explicit rel() form and offers both spellings', () => {
    const e = firstError('rel(used_by, out):(class:server)');
    expect(e.code).toBe('unknown_value');
    expect(e.suggestion).toContain('rel(depends_on, in, …)');
    expect(e.suggestion).toContain('used_by:(…)');
    expect(codes('used_by:(class:server)')).toEqual([]);
  });
});

describe('the suggestions §10 asks for', () => {
  it('names the closest field within edit distance 2', () => {
    expect(firstError('hostnaem:web-1').suggestion).toBe('did you mean "hostname"?');
  });

  it('points a bare namespace at its key form', () => {
    const e = firstError('attr:1');
    expect(e.code).toBe('unknown_field');
    expect(e.suggestion).toContain('attr.<key>');
  });

  it('names where the other namespaces live when nothing is close', () => {
    const e = firstError('completely_unrelated:1');
    expect(e.suggestion).toContain('attr.<name>');
    expect(e.suggestion).toContain('fact.<key>');
  });

  it('sends a numeric band comparison to the numeric twin', () => {
    // §10's worked example. The twin is found through the catalogue — the field
    // sharing the band's column — not hard-coded as a pair.
    expect(firstError('risk >= 70').suggestion).toBe('for the numeric score use "risk_score >= 70"');
  });

  it('sends the retired any-kind identifier spelling to id.any (§13 A1)', () => {
    const e = firstError('id:"aa:bb:cc:dd:ee:ff"');
    expect(e.code).toBe('type_mismatch');
    expect(e.suggestion).toContain('id.any');
    expect(codes('id.any:"aa:bb:cc:dd:ee:ff"')).toEqual([]);
  });

  it('offers the longer enum value a short one is a prefix of', () => {
    // §12 contradiction 2: §9 example 12 writes `status:pending` and the
    // vocabulary is `pending_approval`. Edit distance alone would not find it.
    expect(firstError('status:pending').suggestion).toBe('did you mean "pending_approval"?');
  });
});

describe('errors are ordered and complete', () => {
  it('reports every problem at once, sorted by span', () => {
    const out = check('hostnaem:a and stauts:b', 'asset', cat, opts());
    expect(out.ok).toBe(false);
    if (out.ok) return;
    expect(out.errors.map((e) => e.code)).toEqual(['unknown_field', 'unknown_field']);
    expect(out.errors[0].span.start).toBeLessThan(out.errors[1].span.start);
  });

  it('underlines the offending span', () => {
    const e = firstError('environment:production and hostnaem:web-1');
    expect('environment:production and hostnaem:web-1'.slice(e.span.start, e.span.end)).toBe(
      'hostnaem',
    );
  });
});

File: shared/query/validate/validate_test.go
package validate_test

import (
	"strings"
	"testing"
	"unicode/utf8"

	"github.com/vistasecurity/vistaplatform/shared/query/ast"

	"github.com/vistasecurity/vistaplatform/shared/query/catalog/testcatalog"
	"github.com/vistasecurity/vistaplatform/shared/query/parser"
	"github.com/vistasecurity/vistaplatform/shared/query/queryerr"
	"github.com/vistasecurity/vistaplatform/shared/query/validate"
)

// Every rule in §6 is checked in BOTH polarities: it must reject the thing it
// names and accept the nearest legal query. An over-strict guard is the same
// bug pointed the other way, and this package has one job.

func opts() validate.Options {
	return validate.DefaultOptions().WithLadder(testcatalog.Ladder)
}

// check runs a query and returns its error codes.
func check(t *testing.T, src, target string, o validate.Options) []queryerr.Code {
	t.Helper()
	res, err := parser.Parse(src)
	if err != nil {
		t.Fatalf("Parse(%q): %v", src, err)
	}
	if target == "" {
		target = "asset"
	}
	_, err = validate.Validate(res, target, testcatalog.New(), o)
	if err == nil {
		return nil
	}
	list, ok := err.(queryerr.List)
	if !ok {
		t.Fatalf("Validate(%q) returned %T, want a queryerr.List", src, err)
	}
	return list.Codes()
}

func mustReject(t *testing.T, src string, want queryerr.Code) {
	t.Helper()
	codes := check(t, src, "", opts())
	for _, c := range codes {
		if c == want {
			return
		}
	}
	t.Errorf("Validate(%q) = %v, want %s", src, codes, want)
}

func mustAccept(t *testing.T, src string) {
	t.Helper()
	if codes := check(t, src, "", opts()); codes != nil {
		t.Errorf("Validate(%q) = %v, want no errors", src, codes)
	}
}

func TestUnknownField(t *testing.T) {
	for _, q := range []string{
		"hostnaem:web-1",
		"attr.not_registered:1",
		"fact.not.registered:1",
		"id.not_a_kind:1",
		"attr:1",
		"tag:x and bogus:1",
	} {
		mustReject(t, q, queryerr.CodeUnknownField)
	}
	for _, q := range []string{
		"hostname:web-1",
		"attr.cpu_count:1",
		"fact.os.name:linux",
		"id.serial_number:1",
		"tag.anything:1",
		"tag:x",
	} {
		mustAccept(t, q)
	}
}

// TestUnknownFieldSuggests pins §6's "naming the closest match (edit distance
// ≤ 2)" and §10's worked example.
func TestUnknownFieldSuggests(t *testing.T) {
	res, _ := parser.Parse("environment:production and hostnaem:web-1")
	_, err := validate.Validate(res, "asset", testcatalog.New(), opts())
	list := err.(queryerr.List)
	if len(list) != 1 {
		t.Fatalf("got %d errors, want 1: %v", len(list), list)
	}
	e := list[0]
	if !strings.Contains(e.Suggestion, "hostname") {
		t.Errorf("suggestion %q should name hostname", e.Suggestion)
	}
	// The span must underline the field, not the whole query (§10).
	if got := "environment:production and hostnaem:web-1"[e.Span.Start:e.Span.End]; got != "hostnaem" {
		t.Errorf("span covers %q, want %q", got, "hostnaem")
	}
	// A misspelled relationship is an unknown field, and the suggestion has to
	// come from the relationship vocabulary.
	res, _ = parser.Parse("depands_on:(class:server)")
	_, err = validate.Validate(res, "asset", testcatalog.New(), opts())
	if !strings.Contains(err.(queryerr.List)[0].Suggestion, "depends_on") {
		t.Errorf("suggestion %q should name the depends_on relationship", err.(queryerr.List)[0].Suggestion)
	}
}

func TestOperatorNotAllowed(t *testing.T) {
	for _, q := range []string{
		"environment < production",       // keyword has no ordering
		"class != server",                // class takes '=' only
		`risk_score ~ "7"`,               // no regex on numbers
		"last_seen in (now, now-1d)",     // no in-list on timestamps
		"attr.controller_managed > true", // no ordering on booleans
		"risk_score:web-*",               // no wildcard on numbers
		"class:hard*",                    // no wildcard on a class key
	} {
		mustReject(t, q, queryerr.CodeOperatorNotAllowed)
	}
	for _, q := range []string{
		"environment:production",
		"not class=server",
		"risk_score >= 70",
		"last_seen < now-1d",
		"hostname:web-*",
		"risk_score:[40 to 69]",
	} {
		mustAccept(t, q)
	}
	if codes := check(t, "self_signed:true", "certificate", opts()); codes != nil {
		t.Errorf("a boolean equality is fine, got %v", codes)
	}
}

func TestTypeMismatch(t *testing.T) {
	for _, q := range []string{
		"risk >= 70",         // §10's example: risk is a band
		"risk_score >= high", // and the other way round
		"last_seen < soon",
		"last_seen < now-1mo-15d",
		"segment_id=not-a-uuid",
		"primary_address:999.1.1.1",
		"attr.controller_managed:yes",
		"attr.cpu_count:abc",
	} {
		mustReject(t, q, queryerr.CodeTypeMismatch)
	}
	for _, q := range []string{
		"risk_score >= 70",
		"risk >= high",
		"last_seen < now-45d",
		"segment_id=550e8400-e29b-41d4-a716-446655440000",
		"primary_address:198.51.100.0/24",
		"attr.controller_managed:true",
		"attr.cpu_count:3",
	} {
		mustAccept(t, q)
	}
}

func TestUnknownValue(t *testing.T) {
	for _, q := range []string{
		"status:monitorng",
		"status:pending",
		"class:serverr",
		"ownership:internel",
		"rel(depands_on, out):(class:server)",
		"rel(used_by, out):(class:server)",
	} {
		mustReject(t, q, queryerr.CodeUnknownValue)
	}
	for _, q := range []string{
		"status:monitoring",
		"status:pending_approval",
		"class:server",
		"ownership:internal",
		"rel(depends_on, out):(class:server)",
		"used_by:(class:server)",
	} {
		mustAccept(t, q)
	}
}

// TestNotAssessedNeedsCoverage pins §5.2: not_assessed is only meaningful for a
// band field that records who assessed it.
func TestNotAssessedNeedsCoverage(t *testing.T) {
	mustAccept(t, "risk:not_assessed")
	if codes := check(t, "severity:not_assessed", "finding", opts()); len(codes) == 0 {
		t.Error("severity records no coverage, so not_assessed should be rejected")
	}
	mustAccept(t, "risk:informational")
}

func TestTraversalBudget(t *testing.T) {
	// The default budget is 3 (§11 Q3).
	mustAccept(t, "depends_on(3):(class:server)")
	mustReject(t, "depends_on(4):(class:server)", queryerr.CodeDepthExceeded)

	// The budget is the SUM along the deepest nested path (§5.6).
	mustAccept(t, "depends_on(2):(hosted_on:(class:hypervisor))")
	mustReject(t, "depends_on(2):(hosted_on(2):(class:hypervisor))", queryerr.CodeDepthExceeded)

	// Two traversals side by side each cost their own depth, not the sum.
	mustAccept(t, "depends_on(3):(class:server) and hosted_on(3):(class:hypervisor)")

	// A larger configured budget is honoured up to the hard maximum of 6…
	wide := opts()
	wide.TraversalBudget = 6
	if codes := check(t, "depends_on(6):(class:server)", "", wide); codes != nil {
		t.Errorf("a configured budget of 6 should allow six hops, got %v", codes)
	}
	// …and no configuration may exceed it.
	tooWide := opts()
	tooWide.TraversalBudget = 10
	codes := check(t, "depends_on(7):(class:server)", "", tooWide)
	if len(codes) == 0 || codes[0] != queryerr.CodeDepthExceeded {
		t.Errorf("the hard maximum of 6 must win over configuration, got %v", codes)
	}
	if codes := check(t, "depends_on(6):(class:server)", "", tooWide); codes != nil {
		t.Errorf("six hops is exactly the hard maximum and must pass, got %v", codes)
	}

	// A tightened budget is honoured too.
	tight := opts()
	tight.TraversalBudget = 1
	if codes := check(t, "depends_on(2):(class:server)", "", tight); len(codes) == 0 {
		t.Error("a configured budget of 1 should reject two hops")
	}
	if codes := check(t, "depends_on:(class:server)", "", tight); codes != nil {
		t.Errorf("one hop is within a budget of 1, got %v", codes)
	}
}

func TestRegexRules(t *testing.T) {
	// Compiles as RE2.
	mustAccept(t, `hostname ~ "^web[0-9]+$"`)
	mustAccept(t, `hostname ~ "(?i)^web"`)
	mustReject(t, `hostname ~ "^web[0-9"`, queryerr.CodeRegexInvalid)
	// RE2 has no backreferences by construction, so this is a compile failure
	// rather than a rule of ours.
	mustReject(t, `hostname ~ "(a)\\1"`, queryerr.CodeRegexInvalid)

	// A single repetition count: 255 passes, 256 does not. The bound is
	// Postgres's own (DUPMAX) — §6 said 1000, and every count from 256 up
	// validated and was then refused by the database (§13 A9).
	mustAccept(t, `hostname ~ "a{255}"`)
	mustReject(t, `hostname ~ "a{256}"`, queryerr.CodeRegexInvalid)
	mustAccept(t, `hostname ~ "a{1,255}"`)
	mustReject(t, `hostname ~ "a{1,256}"`, queryerr.CodeRegexInvalid)

	// The PRODUCT across nesting is capped separately, at 1000: two bounds
	// that are each legal can multiply into one that is not, which Postgres
	// reports as "regular expression is too complex" — a 500, where this is a
	// diagnostic with a span.
	mustAccept(t, `hostname ~ "(a{10}){100}"`)
	mustReject(t, `hostname ~ "(a{11}){100}"`, queryerr.CodeRegexInvalid)

	// Length: 256 passes, 257 does not.
	mustAccept(t, `hostname ~ "`+strings.Repeat("a", 256)+`"`)
	mustReject(t, `hostname ~ "`+strings.Repeat("a", 257)+`"`, queryerr.CodeRegexInvalid)
}

// TestRegexLengthIsInRunes is T5. §6 says the regex cap is 256 CHARACTERS and
// the query cap is 4096 BYTES — two different units, in two adjacent rows of
// one table, which is exactly the kind of thing an ASCII-only test cannot tell
// apart. The regex cap was counting bytes, so a pattern of accented characters
// was refused at 128 of them.
//
// Every string here is non-ASCII, so the two units give different answers.
func TestRegexLengthIsInRunes(t *testing.T) {
	const multi = "é" // two bytes, one rune

	at := strings.Repeat(multi, 256)
	if len(at) == utf8.RuneCountInString(at) {
		t.Fatal("test setup: the pattern must be non-ASCII for this to measure anything")
	}
	mustAccept(t, `hostname ~ "`+at+`"`)

	over := strings.Repeat(multi, 257)
	mustReject(t, `hostname ~ "`+over+`"`, queryerr.CodeRegexInvalid)

	// And the message counts the same way it caps.
	codes := check(t, `hostname ~ "`+over+`"`, "asset", opts())
	if len(codes) != 1 || codes[0] != queryerr.CodeRegexInvalid {
		t.Fatalf("got %v, want one regex_invalid", codes)
	}
}

// TestQueryLengthIsInBytes is T5's other half: the query cap stays BYTES, so a
// multibyte query reaches it sooner than its character count suggests. That is
// §6 as written, and the right unit for a size cap on stored text.
func TestQueryLengthIsInBytes(t *testing.T) {
	// "é" is two bytes, so 2048 of them plus the field name is over 4096.
	body := strings.Repeat("é", 2048)
	q := `hostname:"` + body + `"`
	if utf8.RuneCountInString(q) > 4096 {
		t.Fatalf("test setup: %d runes — it must be under the cap in runes and over it in bytes",
			utf8.RuneCountInString(q))
	}
	if len(q) <= 4096 {
		t.Fatalf("test setup: %d bytes, want over 4096", len(q))
	}
	mustReject(t, q, queryerr.CodeQueryTooLong)
}

func TestQueryTooLong(t *testing.T) {
	// The cap is on the source text, at 4096 bytes.
	term := "hostname:aaaaaaaa"
	long := strings.Repeat(term+" or ", 4096/(len(term)+4)+1)
	long = strings.TrimSuffix(long, " or ")
	if len(long) <= 4096 {
		t.Fatalf("test setup: query is only %d bytes", len(long))
	}
	mustReject(t, long, queryerr.CodeQueryTooLong)

	fits := "hostname:" + strings.Repeat("a", 4096-len("hostname:"))
	if len(fits) != 4096 {
		t.Fatalf("test setup: query is %d bytes, want exactly 4096", len(fits))
	}
	mustAccept(t, fits)

	// The failing side, exactly one byte over: a cap tested only with a query
	// far past it cannot tell 4096 from 40960.
	overByOne := fits + "a"
	if len(overByOne) != 4097 {
		t.Fatalf("test setup: query is %d bytes, want exactly 4097", len(overByOne))
	}
	mustReject(t, overByOne, queryerr.CodeQueryTooLong)
}

func TestSizeCaps(t *testing.T) {
	// 64 leaves pass, 65 do not.
	mustAccept(t, strings.TrimSuffix(strings.Repeat("hostname:a and ", 64), " and "))
	mustReject(t, strings.TrimSuffix(strings.Repeat("hostname:a and ", 65), " and "),
		queryerr.CodeTooManyClauses)

	// 16 sub-predicates pass, 17 do not.
	mustAccept(t, strings.TrimSuffix(strings.Repeat("endpoint:(port:1) and ", 16), " and "))
	mustReject(t, strings.TrimSuffix(strings.Repeat("endpoint:(port:1) and ", 17), " and "),
		queryerr.CodeTooManyClauses)

	// 16 levels of parentheses pass, 17 do not.
	mustAccept(t, strings.Repeat("(", 16)+"hostname:a"+strings.Repeat(")", 16))
	mustReject(t, strings.Repeat("(", 17)+"hostname:a"+strings.Repeat(")", 17),
		queryerr.CodeTooManyClauses)

	// 256 values in one list pass, 257 do not. The field is business_unit
	// rather than environment because the cap is what is under test: the values
	// are filler, and a closed enum would fail them for a second reason.
	mustAccept(t, "business_unit in ("+strings.TrimSuffix(strings.Repeat("a, ", 256), ", ")+")")
	mustReject(t, "business_unit in ("+strings.TrimSuffix(strings.Repeat("a, ", 257), ", ")+")",
		queryerr.CodeTooManyClauses)
}

func TestUntranslatable(t *testing.T) {
	// A collection with no shape from this target.
	if codes := check(t, "finding:(kind:stale)", "endpoint", opts()); len(codes) == 0 ||
		codes[0] != queryerr.CodeUntranslatable {
		t.Errorf("finding from an endpoint should be untranslatable, got %v", codes)
	}
	if codes := check(t, "asset:(class:server)", "endpoint", opts()); codes != nil {
		t.Errorf("asset from an endpoint is reachable, got %v", codes)
	}
	// Free text where §5.4's column set does not exist.
	if codes := check(t, "payroll", "certificate", opts()); len(codes) == 0 {
		t.Error("free text on a certificate should be untranslatable")
	}
	if codes := check(t, "subject_dn:payroll", "certificate", opts()); codes != nil {
		t.Errorf("a named field on a certificate is fine, got %v", codes)
	}
	// Traversal from a target that is not an asset.
	if codes := check(t, "depends_on:(class:server)", "certificate", opts()); len(codes) == 0 {
		t.Error("traversal from a certificate should be untranslatable")
	}
}

// TestBandWithoutLadder is the fail-closed case: without a ladder there is no
// way to compare a band, and guessing a threshold is exactly what §5.5 forbids.
func TestBandWithoutLadder(t *testing.T) {
	res, _ := parser.Parse("risk >= high")
	_, err := validate.Validate(res, "asset", testcatalog.New(), validate.DefaultOptions())
	if err == nil {
		t.Fatal("a band comparison without a ladder must be refused")
	}
	if !err.(queryerr.List).Has(queryerr.CodeUntranslatable) {
		t.Errorf("got %v, want untranslatable", err)
	}
}

func TestUnknownTarget(t *testing.T) {
	res, _ := parser.Parse("hostname:web-1")
	_, err := validate.Validate(res, "nonesuch", testcatalog.New(), opts())
	if err == nil {
		t.Fatal("an unknown target must be refused")
	}
}

func TestEmptyQueryIsValid(t *testing.T) {
	mustAccept(t, "")
}

// TestResolutionFillsTheFieldRef proves the validator is what makes a query
// translatable: before it runs, no field has a type or an accessor.
func TestResolutionFillsTheFieldRef(t *testing.T) {
	res, _ := parser.Parse("hostname:web-1")
	root, err := validate.Validate(res, "asset", testcatalog.New(), opts())
	if err != nil {
		t.Fatal(err)
	}
	seen := 0
	ast.Fields(root, func(f *ast.FieldRef) {
		seen++
		if !f.Resolved {
			t.Errorf("field %q was not resolved", f.Text)
		}
		if f.Type != ast.TypeText || f.Accessor.Column != "hostname" {
			t.Errorf("field %q resolved to %s/%q", f.Text, f.Type, f.Accessor.Column)
		}
	})
	if seen != 1 {
		t.Errorf("visited %d fields, want 1", seen)
	}
}
